lua_examples/luaOTA/check.lua

```diff
@@ -0,0 +1,66 @@
+--SAFETRIM
+--------------------------------------------------------------------------------
+-- LuaOTA provisioning system for ESPs using NodeMCU Lua
+-- LICENCE: http://opensource.org/licenses/MIT
+-- TerryE  15 Jul 2017
+--
+-- See luaOTA.md for description and implementation notes
+
+--------------------------------------------------------------------------------
+
+-- upvals
+local crypto, file, json,  net, node, table, tmr, wifi =
+      crypto, file, sjson, net, node, table, tmr, wifi
+local error, pcall   = error, pcall 
+local loadfile, gc   = loadfile, collectgarbage
+local concat, unpack = table.concat, unpack or table.unpack 
+
+local self = {post = node.task.post, prefix = "luaOTA/", conf = {}}
+
+self.log     = (DEBUG == true) and print or function() end 
+self.modname = ...  
+
+--------------------------------------------------------------------------------------
+-- Utility Functions
+
+setmetatable( self, {__index=function(self, func) --upval: loadfile
+    -- The only __index calls in in LuaOTA are dynamically loaded functions. 
+    -- The convention is that functions starting with "_" are treated as 
+    -- call-once / ephemeral; the rest are registered in self 
+    func = self.prefix .. func
+    local f,msg = loadfile( func..".lc")
+    if msg then f, msg = loadfile(func..".lua") end
+    if msg then error (msg,2) end
+    if func:sub(8,8) ~= "_" then self[func] = f end
+    return f
+  end} )
+
+function self.sign(arg)  --upval: crypto, json, self
+  arg = json.encode(arg)
+  return arg .. crypto.toHex(crypto.hash("MD5", self.secret .. arg):sub(-3)) .. '\n'
```

Comment:
To perform a HMAC-MD5 with no additional packages, it suffices to use
```
function hmd5(key, data)
  if key:len() > 64 then key = md5.sum(key) end
  key = key .. string.rep('\0',64-key:len())
  local ki = md5.exor(string.rep(string.char(0x36),64),key)
  local ko = md5.exor(string.rep(string.char(0x5c),64),key)
  return md5.sumhexa(ko..md5.sum(ki..data))
end
```

This function passes all the RFC2022 (https://tools.ietf.org/html/rfc2202) test vectors, as can be verified by running

```
print(hmd5(string.rep(string.char(0x0B),16), "Hi There") == "9294727a3638bb1c13f48ef8158bfc9d")
print(hmd5("Jefe", "what do ya want for nothing?") == "750c783e6ab0b503eaa86e310a5db738")
print(hmd5(string.rep(string.char(0xaa),16), string.rep(string.char(0xdd),50)) == "56be34521d144c88dbb8c733f0e8b3f6")
print(hmd5("\1\2\3\4\5\6\7\8\9\10\11\12\13\14\15\16\17\18\19\20\21\22\23\24\25",string.rep(string.char(0xcd),50)) == "697eaf0aca3a3aea3a75164746ffaa79")
print(hmd5(string.rep(string.char(0x0c),16),"Test With Truncation") == "56461ef2342edc00f9bab995690efd4c")
print(hmd5(string.rep(string.char(0xaa),80),"Test Using Larger Than Block-Size Key - Hash Key First") == "6b1ab7fe4bd7bf8f0b62e6ce61b9d0cd")
print(hmd5(string.rep(string.char(0xaa),80),"Test Using Larger Than Block-Size Key and Larger Than One Block-Size Data") == "6f630fad67cda0ee1fb1f562db3aa53e")
```